File: app/src/main/java/ch/unstable/ost/FavoriteCardFragment.java
package ch.unstable.ost;

import android.arch.persistence.room.EmptyResultSetException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import ch.unstable.ost.api.model.Connection;
import ch.unstable.ost.database.Databases;
import ch.unstable.ost.database.dao.FavoriteConnectionDao;
import ch.unstable.ost.database.model.FavoriteConnection;
import ch.unstable.ost.lists.query.QueryBinder;
import ch.unstable.ost.utils.LocalizationUtils;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class FavoriteCardFragment extends Fragment {

    private static final String TAG = "FavoriteCardFragment";
    private OnFavoriteSelectedListener mListener;
    private View mCardFavorites;
    private TextView mFavoriteFromTo;
    private TextView mFavoriteDate;
    private FavoriteConnectionDao mFavoriteDao;
    @Nullable
    private FavoriteConnection mLatestFavorite = null;
    private Disposable mDisposable;

    public FavoriteCardFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mFavoriteDao = Databases.getCacheDatabase(getContext()).favoriteConnectionDao();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.card_favorites, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mCardFavorites = view.findViewById(R.id.cardFavorites);
        mCardFavorites.setVisibility(View.GONE);
        mFavoriteFromTo = view.findViewById(R.id.favoriteFromTo);
        mFavoriteDate = view.findViewById(R.id.favoriteDate);

        View moreButton = view.findViewById(R.id.buttonMoreFavorites);
        moreButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onMoreButtonPressed();
            }
        });
        View openButton = view.findViewById(R.id.buttonOpenFavorite);
        openButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onOpenButtonPressed();
            }
        });
    }

    private void bindConnection(FavoriteConnection favoriteConnection) {
        Connection connection = favoriteConnection.getConnection();
        QueryBinder.bindFromToText(mFavoriteFromTo, connection);
        mFavoriteDate.setText(LocalizationUtils.getDepartureText(getContext(), connection.getDepartureDate()));
        mCardFavorites.setVisibility(View.VISIBLE);
        mLatestFavorite = favoriteConnection;
    }

    @Override
    public void onResume() {
        Log.d(TAG, "onResume()");
        super.onResume();
        if(mDisposable != null) mDisposable.dispose();
        mDisposable = mFavoriteDao.getLatestFavorite()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<FavoriteConnection>() {
                    @Override
                    public void accept(FavoriteConnection favoriteConnection) throws Exception {
                        Log.d(TAG, "Got latest connection: " + favoriteConnection);
                        bindConnection(favoriteConnection);
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        mCardFavorites.setVisibility(View.GONE); // TODO: animate
                        if(throwable instanceof EmptyResultSetException) {
                            Log.d(TAG, "No last favorite");
                        } else {
                            Log.e(TAG, "Failed to load latest favorite: ", throwable);
                            // TODO: error handling
                        }
                    }
                });
    }

    @Override
    public void onStop() {
        super.onStop();
        mDisposable.dispose();
    }

    private void onOpenButtonPressed() {
        if(mListener != null && mLatestFavorite != null) {
            mListener.onFavoriteSelected(mLatestFavorite);
        } else if(BuildConfig.DEBUG) {
            Log.w(TAG, "mListener or mLatestFavorite is null (" + mListener + ", " + mLatestFavorite + ")");
        }
    }

    private void onMoreButtonPressed() {
        Intent intent = new Intent(getContext(), FavoritesActivity.class);
        startActivity(intent);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFavoriteSelectedListener) {
            mListener = (OnFavoriteSelectedListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFavoriteSelectedListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public interface OnFavoriteSelectedListener {
        void onFavoriteSelected(@NonNull FavoriteConnection favoriteConnection);
    }
}

File: app/src/main/java/ch/unstable/ost/FavoriteCardFragment.kt
package ch.unstable.ost

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import ch.unstable.ost.database.Databases
import ch.unstable.ost.database.dao.FavoriteConnectionDao
import ch.unstable.ost.database.model.FavoriteConnection
import ch.unstable.ost.utils.LocalizationUtils.getDepartureText
import ch.unstable.ost.views.lists.query.QueryBinder.bindFromToText
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers

class FavoriteCardFragment : QuickstartCardFragment() {
    private var favoriteSelectedListener: OnFavoriteSelectedListener? = null
    private var cardFavorites: View? = null
    private var favoriteFromTo: TextView? = null
    private var favoriteDate: TextView? = null
    private var favoriteDao: FavoriteConnectionDao? = null
    private var latestFavorite: FavoriteConnection? = null
    private var disposable: Disposable? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        favoriteDao = Databases.getCacheDatabase(requireContext()).favoriteConnectionDao()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.card_favorites, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        cardFavorites = view.findViewById(R.id.cardFavorites)
        cardFavorites?.visibility = View.GONE
        favoriteFromTo = view.findViewById(R.id.favoriteFromTo)
        favoriteDate = view.findViewById(R.id.favoriteDate)
        val moreButton = view.findViewById<View>(R.id.buttonMoreFavorites)
        moreButton.setOnClickListener { onMoreButtonPressed() }
        val openButton = view.findViewById<View>(R.id.buttonOpenFavorite)
        openButton.setOnClickListener { onOpenButtonPressed() }
    }

    private fun bindConnection(favoriteConnection: FavoriteConnection) {
        val connection = favoriteConnection.connection
        bindFromToText(favoriteFromTo!!, connection)
        favoriteDate!!.text = getDepartureText(requireContext(), connection.departureDate)
        cardFavorites!!.visibility = View.VISIBLE
        latestFavorite = favoriteConnection
    }

    override fun onResume() {
        Log.d(TAG, "onResume()")
        super.onResume()
        disposable?.dispose()
        disposable = favoriteDao!!.getLatestFavorite()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(Consumer { favoriteConnection: FavoriteConnection ->
                    Log.d(TAG, "Got latest connection: $favoriteConnection")
                    bindConnection(favoriteConnection)
                }, errorConsumer)
    }

    override fun onStop() {
        super.onStop()
        disposable?.dispose()
    }

    private fun onOpenButtonPressed() {
        if (favoriteSelectedListener != null && latestFavorite != null) {
            favoriteSelectedListener!!.onFavoriteSelected(latestFavorite!!)
        } else if (BuildConfig.DEBUG) {
            Log.w(TAG, "mListener or mLatestFavorite is null ($favoriteSelectedListener, $latestFavorite)")
        }
    }

    private fun onMoreButtonPressed() {
        val intent = Intent(context, FavoritesActivity::class.java)
        startActivity(intent)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        favoriteSelectedListener = if (context is OnFavoriteSelectedListener) {
            context
        } else {
            throw RuntimeException(context.toString()
                    + " must implement OnFavoriteSelectedListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        favoriteSelectedListener = null
    }

    override val errorMessage: Int
        get() = R.string.error_failed_to_load_favorites

    interface OnFavoriteSelectedListener {
        fun onFavoriteSelected(favoriteConnection: FavoriteConnection)
    }

    companion object {
        private const val TAG = "FavoriteCardFragment"
    }
}